src/gennylib/include/gennylib/PhaseLoop.hpp

```diff
@@ -149,6 +162,16 @@ class IterationChecker final {
         return _doesBlock;
     }
 
+    void sleepBefore(const Orchestrator& o, const PhaseNumber pn) const {
+        if (_sleeper)
+            _sleeper->before(o, pn);
+    }
```

Comment:
no strong opinion on this but consider just always calling `_sleeper->before()` (and `after()`) and just let the sleeper itself decide if it wants to sleep.

I realize this is a change also to not use `unique_ptr` and to potentially use something like `std::optional` in `Sleeper` (or skip the `sleep_for` if the duration is zero).

This way the sleeper owns all the logic about when to sleep and how much etc. E.g. a change could conceivably be made to alter the sleep by coordinating more closely with the Orchestrator or something.